contracts/nois-airdrop/src/contract.rs

```diff
@@ -0,0 +1,790 @@
+        },
+    )?;
+
+    let response = Response::new().add_message(WasmMsg::Execute {
+        contract_addr: nois_proxy.into(),
+        // GetRandomnessAfter requests the randomness from the proxy after a specific timestamp
+        // The job id is needed to know what randomness we are referring to upon reception in the callback.
+        // In this example we only need 1 random number so this can be hardcoded to "airdrop"
+        msg: to_binary(&ProxyExecuteMsg::GetRandomnessAfter {
+            after: random_beacon_after,
+            job_id: "airdrop".to_string(),
+        })?,
+        // We pay here the proxy contract with whatever the manager sends. The manager needs to check in advance the proxy prices.
+        funds: info.funds, // Just pass on all funds we got
+    });
+    Ok(response)
+}
+
+pub fn execute_receive(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    callback: NoisCallback,
+) -> Result<Response, ContractError> {
+    let RandomnessParams {
+        nois_proxy,
+        nois_randomness: _,
+        requested,
+    } = NOIS_RANDOMNESS.load(deps.storage).unwrap();
+
+    // callback should only be allowed to be called by the proxy contract
+    // otherwise anyone can cut the randomness workflow and cheat the randomness by sending the randomness directly to this contract
+    ensure_eq!(info.sender, nois_proxy, ContractError::UnauthorizedReceive);
+    let randomness: [u8; 32] = callback
+        .randomness
+        .to_array()
+        .map_err(|_| ContractError::InvalidRandomness)?;
+    // Make sure the randomness does not exist yet
+
+    match NOIS_RANDOMNESS.load(deps.storage).unwrap().nois_randomness {
+        None => NOIS_RANDOMNESS.save(
+            deps.storage,
+            &RandomnessParams {
+                nois_proxy,
+                nois_randomness: Some(randomness),
+                requested,
+            },
+        ),
+        Some(_randomness) => return Err(ContractError::ImmutableRandomness),
+    }?;
+
+    Ok(Response::default())
+}
+
+fn execute_withdraw_all(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    // check the calling address is the authorised addr
+
+    // if manager set validate, otherwise unauthorized
+    let manager = config.manager.ok_or(ContractError::Unauthorized {})?;
+    if info.sender != manager {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let amount = deps
+        .querier
+        .query_balance(env.contract.address, AIRDROP_DENOM)?;
+    let msg = BankMsg::Send {
+        to_address: config.withdrawal_address.into(),
+        amount: vec![amount],
+    };
+    let res = Response::new()
+        .add_message(msg)
+        .add_attribute("action", "withdraw_all");
+    Ok(res)
+}
+
+fn is_randomly_eligible(addr: &Addr, randomness: [u8; 32]) -> bool {
+    let mut hasher = Sha256::new();
+    hasher.update(addr.as_bytes());
```

Comment:
```suggestion
    hasher.update(addr.as_bytes());
    hasher.update(randomness);
```

you can hash them together in one hasher. The second ones is not needed.